### PR TITLE
Delivery via `Inbox`

### PR DESCRIPTION
**Pull request by armiol**
Before this PR the signals such as `Event`s and `Command`s were being delivered to the entities directly via the message endpoints. Such a way caused the issues in a multi-threaded (single node) or multi-node environment. The same entity could have been updated simultaneously from several execution threads, which in turn caused the loss of data.

In this PR an entity `Inbox` along with the application-wide `Delivery` sub-system were introduced.

### Inbox

Each entity repository now operates via the `Inbox`. Every message dispatched to one of the entities gets to the inbox of the respective entity. In it, the messages are split into shards and are processed on an "one-processor => one-shard" basis. A server-wide `Delivery` mechanism is responsible for serving the messages from shards to the entities and for preventing the simultaneous access the same entity from the multiple threads and-or application server nodes.

`Inbox` is a part of an internal framework API and no specific configuration is required in order to operate with it.

### Delivery

**Highlights**

* splits the incoming messages into shards and allows to deliver messages to their destinations on a per-shard basis;
* guarantees that one and only one application server node serves the messages from the given shard at a time, thus preventing any concurrent modifications of entity state.
* de-duplicates the messages per-target;

**Usage**

The delivery mechanism is single per-JVM and resides in the `ServerEnvironment` singleton.

The framework users may obtain the current delivery via 

```java
Delivery delivery = ServerEnvironment.instance()
                                     .delivery();
```                                  
By default, the delivery is set to "local", based upon in-memory structures and suitable for running the unit tests. See `Delivery.local()` for the implementation details.

The framework users may update the current delivery by using          
```java
ServerEnvironment.instance()
                 .configureDelivery(myDelivery);
```
This should be done in real-world environments upon the application start. It is NOT recommended to swap the delivery implementation at runtime.

**Configuration**

*Shard Assignment*

Once a message arrives at the inbox, a target shard is assigned to it. The shard index is built on top of the identifier of the target entity. Therefore the messages heading to a particular entity always reside in the same shard.

The framework users may customize this behaviour via

```java
Delivery.newBuilder() 
        .setStrategy(myStrategy)
        // the rest of the configuration.
```

The default strategy uses the hash code of an entity ID as a basis of determining the shard index. The index is determined as `id.hashCode() % <total number of shards>`.

*Idempotence Window*

During the delivery the messages delivered to their targets are de-duplicated. As it is hardly possible to maintain the whole history of previously delivered messages, the framework keeps track only of the messages delivered recently. This time window is configurable:

```java
Delivery.newBuilder() 
        .setIdempotenceWindow(Durations.fromMillis(500))
        // the rest of the configuration.
```

By default, the "local" delivery strategy uses a wide window of 30 seconds. Such a value allows to run any tests and be sure there will be no duplicates delivered.

*Scaling*

In the real multi-node environment the framework users would want to scale their instances upon the arrival of new messages to each of the shards. In order to do so they should register an observer, which will be notified of the messages written to their shards:

```java
ServerEnvironment.instance()
                 .delivery()
                 .subscribe(myObserver);
```

Depending on the environment, the framework users might want to use a messaging queue to broadcast the index of the shard which requires processing. 

```java
public class MyObserver implements ShardObserver {

    // ...

    @Override
    public void onMessage(InboxMessage update) {
        queue.push(update.getShardIndex());
    }
}
```
The node which receives such a message will then pick up the shard and starts to deliver the messages to their targets.

```java

// called when the queue dispatches the message
void callback(ShardIndex index) {
    Delivery delivery = ServerEnvironment.instance()
                                         .delivery();
    delivery.deliverMessagesFrom(index);
}
```

The local version of the delivery uses synchronous dispatching mechanism which is triggered for each observed message.

*Storage*

To operate the delivery mechanism relies on an application-wide `InboxStorage`. It should be supplied via the builder:

```java
Delivery.newBuilder() 
        .setInboxStorage(myStorage)
        // the rest of the configuration.
```
Each of the per-repository inboxes will share this instance of the `InboxStorage` — as each of `Inbox` instances registers itself in the current `Delivery` instance automatically.

It is important that `InboxStorage` is not like other types of storages produced by `StorageFactory`. `StorageFactory` allows creating `BoundedContext`-specific storages, while `InboxStorage` is wider and covers multiple bounded contexts. It is recommended (but not required) to always have it multi-tenant, in case more bounded contexts are added to the application, and some of them turn out to be multi-tenant.

*Units of Work*

To maintain "one-processor => one-shard" policy a special work registry should be configured for the `Delivery`:

```java
Delivery.newBuilder() 
        .setWorkRegistry(myRegistry)
        // the rest of the configuration.
```

The registry is responsible for keeping track of which shards are served by which nodes. In case the shard is already taken by some node, the work registry does not allow to pick the shard again. The simplest way to implement such a registry in the real-world environments is to base it upon strongly-consistent storage.

### What's next

In the scope of this feature a few issues were created to be addressed later:

* #1094 — emit events instead of exceptions. 
* #1095 — allow to configure the node ID.

Also, the implementation of command scheduling and projection catch-up will become possible on top of this PR. The features will be implemented in future PRs.

### CODE REVIEW

File: server/src/main/java/io/spine/server/delivery/Delivery.java [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+
+            maybePage = currentPage.next();
+        }
+        return totalMessagesDelivered;
+    }
+
+    private ImmutableList<RuntimeException>
+    deliverByType(ImmutableList<InboxMessage> toDeliver, ImmutableList<InboxMessage> idmptSource) {
+
+        Map<String, List<InboxMessage>> messagesByType = groupByTargetType(toDeliver);
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/delivery/Delivery.java]: Maybe a `Multimap` would look better.

File: server/src/main/java/io/spine/server/delivery/Delivery.java [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+     * Subscribes to the updates of shard contents.
+     *
+     * <p>The passed observer will be notified that the contents of a shard with a particular index
+     * were changed.
+     *
+     * @param observer
+     *         an observer to notify of updates.
+     */
+    public void subscribe(ShardObserver observer) {
+        shardObservers.add(observer);
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/delivery/Delivery.java]: Please add a null check.

File: server/src/main/java/io/spine/server/delivery/DeliveryStrategy.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+     * @return the shard index
+     */
+    ShardIndex getIndexFor(Object entityId);
+
+    /**
+     * Tells how many shards there are according to this strategy.
+     *
+     * @return total count of shards
+     */
+    int getShardCount();
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/delivery/DeliveryStrategy.java]: Please remove `get` prefixes in this type.

File: server/src/main/java/io/spine/server/delivery/InboxOfCommands.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        return CommandEnvelope.of(message.getCommand());
+    }
+
+    @Override
+    protected Dispatcher dispatcherWith(Collection<InboxMessage> deduplicationSource) {
+        return new CommandDispatcher(deduplicationSource);
+    }
+
+    @Override
+    protected InboxMessageStatus getStatus(CommandEnvelope message) {
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/delivery/InboxOfCommands.java]: Please remove `get`.

File: server/src/main/java/io/spine/server/delivery/InboxOfEvents.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        private EventDispatcher(Collection<InboxMessage> deduplicationSource) {
+            super(deduplicationSource);
+        }
+
+        @Override
+        protected Predicate<? super InboxMessage> filterByType() {
+            return (Predicate<InboxMessage>) InboxMessage::hasEvent;
+        }
+
+
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/delivery/InboxOfEvents.java]: Please remove this empty line.

File: server/src/main/java/io/spine/server/delivery/LocalDispatchingObserver.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ */
+public class LocalDispatchingObserver implements ShardObserver {
+
+    @Override
+    public void onMessage(InboxMessage update) {
+        Delivery delivery = ServerEnvironment.instance()
+                                             .delivery();
+        delivery.deliverMessagesFrom(update.getShardIndex());
+    }
+}
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/delivery/LocalDispatchingObserver.java]: Please add the EOF. It seems to be a bug in IDEA. I've already seen this with my own changeset as well.

File: server/src/main/java/io/spine/server/delivery/ShardedStorage.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+     * Reads the contents of the storage by the given shard index and returns the first page
+     * of the results.
+     *
+     * <p>The older items go first.
+     *
+     * @param index
+     *         the shard index to return the results for
+     * @return the first page of the results
+     */
+    public abstract Page<M> contentsBackwards(ShardIndex index);
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/delivery/ShardedStorage.java]: Maybe, history-backwards?
It is rather history-forwards though.

Comment (dmdashenkov) [reply server/src/main/java/io/spine/server/delivery/ShardedStorage.java]: Maybe something like `chronologically`?

File: server/src/main/java/io/spine/server/delivery/ShardedStorage.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+import io.spine.server.storage.AbstractStorage;
+import io.spine.server.storage.ReadRequest;
+
+/**
+ * An abstract class for the {@link io.spine.server.storage.RecordStorage RecordStorage}s, which
+ * spread their records across shards.
+ *
+ * @param <M>
+ *         the type of the messages stored
+ * @author Alex Tymchenko
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/delivery/ShardedStorage.java]: I thought we've agreed not to use `@author` tags anymore.

File: server/src/main/java/io/spine/server/delivery/memory/InMemoryShardedWorkRegistry.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        workByNode.put(index, record);
+        return Optional.of(asSession(record));
+    }
+
+    @SuppressWarnings("ResultOfMethodCallIgnored")      // `Builder` methods called in `if-else`.
+    @CanIgnoreReturnValue
+    private ShardSessionRecord updatePickedBy(ShardSessionRecord record,
+                                              @Nullable NodeId nodeId) {
+        ShardSessionRecord.Builder builder = record.toBuilder();
+        if(nodeId == null) {
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/delivery/memory/InMemoryShardedWorkRegistry.java]: Please fix formatting here.

File: server/src/main/java/io/spine/server/projection/ProjectionRepository.java [hunk 10]
```diff
@@ -326,6 +346,15 @@ EventStreamQuery createStreamQuery() {
         return builder.build();
     }
 
+    @OverridingMethodsMustInvokeSuper
+    @Override
+    public void close() {
+        super.close();
+        if(inbox != null) {
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/projection/ProjectionRepository.java]: Please fix formatting

File: server/src/test/java/io/spine/server/delivery/given/package-info.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+ * Test environment classes for testing {@code io.spine.server.delivery} package.
+ */
+
+@CheckReturnValue
+@ParametersAreNonnullByDefault
+package io.spine.server.delivery.given;
+
+import com.google.errorprone.annotations.CheckReturnValue;
+
+import javax.annotation.ParametersAreNonnullByDefault;
```
Comment (dmdashenkov) [inline server/src/test/java/io/spine/server/delivery/given/package-info.java]: Please add an EOF.

File: server/src/test/proto/spine/test/delivery/delivery_commands.proto [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+option java_outer_classname = "DeliveryTestCommandsProto";
+option java_multiple_files = true;
+
+message AddNumber {
+    option (is).java_type = "io.spine.server.delivery.given.CalculatorSignal";
+
+    string calculator_id = 1;
+
+    int32 value = 2;
+}
```
Comment (dmdashenkov) [inline server/src/test/proto/spine/test/delivery/delivery_commands.proto]: Here as well.

File: server/src/test/proto/spine/test/delivery/delivery_events.proto [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    int32 value = 2;
+}
+
+message NumberReacted {
+    option (is).java_type = "io.spine.server.delivery.given.CalculatorSignal";
+
+    string calculator_id = 1;
+
+    int32 value = 2;
+}
```
Comment (dmdashenkov) [inline server/src/test/proto/spine/test/delivery/delivery_events.proto]: Also here.

File: server/src/main/java/io/spine/server/ServerEnvironment.java [hunk 14]
```diff
@@ -49,27 +51,51 @@
...
+                       .setAppId(applicationId)
+                       .setValue(Identifier.newUuid())
+                       .vBuild();
     }
 
     /**
      * Returns a singleton instance.
      */
-    public static ServerEnvironment getInstance() {
+    public static ServerEnvironment instance() {
```
[the remark this answers is not in the document]
Comment (armiol) [reply server/src/main/java/io/spine/server/ServerEnvironment.java]: Done.

File: server/src/main/java/io/spine/server/aggregate/AggregateRepository.java [hunk 15]
```diff
@@ -590,4 +596,13 @@ protected EntityLifecycle lifecycleOf(I id) {
     private void onCommandTargetSet(I id, CommandId commandId) {
         lifecycleOf(id).onTargetAssignedToCommand(commandId);
     }
+
+    @OverridingMethodsMustInvokeSuper
+    @Override
+    public void close() {
+        super.close();
+        if(inbox != null) {
```
[the remark this answers is not in the document]
Comment (armiol) [reply server/src/main/java/io/spine/server/aggregate/AggregateRepository.java]: Done.

File: server/src/main/java/io/spine/server/delivery/Delivery.java [hunk 16]
```diff
@@ -0,0 +1,481 @@
...
+ */
+public final class Delivery {
+
+    /**
+     * The width of the idempotence window in a local environment.
+     *
+     * <p>Selected to be pretty big to avoid dispatching duplicates to any entities.
+     */
+    private static final Duration LOCAL_IDEMPOTENCE_WINDOW = Durations.fromSeconds(30);
+    /**
```
[the remark this answers is not in the document]
Comment (armiol) [reply server/src/main/java/io/spine/server/delivery/Delivery.java]: Done.

File: server/src/main/java/io/spine/server/delivery/Delivery.java [hunk 17]
```diff
@@ -0,0 +1,481 @@
...
+            Integer deliveredMsgCount = null;
+            while (deliveredMsgCount == null || deliveredMsgCount > 0) {
+                deliveredMsgCount = doDeliver(session);
+            }
+        } finally {
+            session.complete();
+        }
+    }
+
+    private int doDeliver(ShardProcessingSession session) {
```
[the remark this answers is not in the document]
Comment (armiol) [reply server/src/main/java/io/spine/server/delivery/Delivery.java]: Not sure it helped.

[the remark this answers is not in the document]
Comment (armiol) [reply server/src/main/java/io/spine/server/delivery/Delivery.java]: Done.

File: server/src/main/java/io/spine/server/delivery/Delivery.java (hunk 1 above)
Comment (armiol) [reply server/src/main/java/io/spine/server/delivery/Delivery.java]: Let's discuss, but I don't think so.

File: server/src/main/java/io/spine/server/delivery/Delivery.java (hunk 2 above)
Comment (armiol) [reply server/src/main/java/io/spine/server/delivery/Delivery.java]: Not sure why would we. Let's discuss.

File: server/src/main/java/io/spine/server/projection/ProjectionRepository.java (hunk 10 above)
Comment (armiol) [reply server/src/main/java/io/spine/server/projection/ProjectionRepository.java]: Done.

**Comment by armiol** [comment]
@dmdashenkov most of the comments addressed. The rest is up for the vocal discussion.

File: server/src/main/java/io/spine/server/delivery/Delivery.java [hunk 18]
```diff
@@ -0,0 +1,496 @@
...
+        Page<InboxMessage> startingPage = inboxStorage.readAll(index);
+        Optional<Page<InboxMessage>> maybePage = Optional.of(startingPage);
+
+        int totalMessagesDelivered = 0;
+        while (maybePage.isPresent()) {
+            Page<InboxMessage> currentPage = maybePage.get();
+            ImmutableList<InboxMessage> messages = currentPage.contents();
+            if (messages.isEmpty()) {
+                maybePage = currentPage.next();
+                continue;
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/delivery/Delivery.java]: We could get rid of this `continue` instruction. Just invert the condition and extract
```java
maybePage = currentPage.next();
```
out of the `if` statement, as both branches end with this statement.

File: server/src/main/java/io/spine/server/delivery/ShardProcessingSession.java [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+import io.spine.annotation.SPI;
+import io.spine.server.NodeId;
+
+/**
+ * The session of processing the messages, which reside in a shard.
+ *
+ * <p>Starts by {@linkplain ShardedWorkRegistry#pickUp(ShardIndex, NodeId)} picking up}
+ * the shard to process.
+ *
+ * @author Alex Tymchenko
```
Comment (alexander-yevsyukov) [inline server/src/main/java/io/spine/server/delivery/ShardProcessingSession.java]: Let's not use author tags, as we agreed recently.

File: server/src/main/java/io/spine/server/delivery/ShardedMessageDelivery.java [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+package io.spine.server.delivery;
+
+import java.util.List;
+
+/**
+ * A delivery routine of postponed processing the previously sharded messages.
+ *
+ * @param <M>
+ *         the type of sharded messages
+ * @author Alex Tymchenko
```
Comment (alexander-yevsyukov) [inline server/src/main/java/io/spine/server/delivery/ShardedMessageDelivery.java]: Let's not use author tags, as we agreed recently.

File: server/src/main/java/io/spine/server/delivery/ShardedRecord.java [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+
+import com.google.protobuf.Message;
+import com.google.protobuf.Timestamp;
+import io.spine.annotation.GeneratedMixin;
+import io.spine.annotation.Internal;
+
+/**
+ * A Protobuf {@link com.google.protobuf.Message Message} used for storage in a sharded environment.
+ *
+ * @author Alex Tymchenko
```
Comment (alexander-yevsyukov) [inline server/src/main/java/io/spine/server/delivery/ShardedRecord.java]: Let's not use author tags, as we agreed recently.

File: server/src/main/java/io/spine/server/delivery/ShardedWorkRegistry.java [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+import io.spine.annotation.SPI;
+import io.spine.server.NodeId;
+
+import java.util.Optional;
+
+/**
+ * The registry of the shard indexes along with the identifiers of the nodes, which
+ * process the messages corresponding to each index.
+ *
+ * @author Alex Tymchenko
```
Comment (alexander-yevsyukov) [inline server/src/main/java/io/spine/server/delivery/ShardedWorkRegistry.java]: Let's not use author tags, as we agreed recently.

File: server/src/main/proto/spine/server/delivery/delivery.proto [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+
+import "google/protobuf/timestamp.proto";
+
+import "spine/server/server_environment.proto";
+
+// An index of a shard.
+//
+// A value type used across the application. To be potentially used in JavaScript, Go, C++ and
+// other client and server environments, that are split into shards.
+message ShardIndex {
```
Comment (alexander-yevsyukov) [inline server/src/main/proto/spine/server/delivery/delivery.proto]: Let's add empty line if the preceding comment has more than one paragraph. And _not_ have an empty line if the comment is one paragraph only.

File: server/src/main/proto/spine/server/delivery/delivery.proto [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+    // The zero-based index of the shard.
+    int32 index = 1 [(min).value = "0"];
+
+    // Total number of shards.
+    int32 of_total = 2 [(min).value = "1"];
+}
+
+
+// A record describing the session in scope of which sharded messages are being processed.
+//
```
Comment (alexander-yevsyukov) [inline server/src/main/proto/spine/server/delivery/delivery.proto]: Please remove the empty line here.

File: server/src/main/proto/spine/server/delivery/delivery.proto [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+message ShardSessionRecord {
+
+    // The index of a shard processed in this session.
+    //
```
Comment (alexander-yevsyukov) [inline server/src/main/proto/spine/server/delivery/delivery.proto]: Please remove the empty line.

File: server/src/main/proto/spine/server/delivery/delivery.proto [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+    ShardIndex index = 1 [(required) = true];
+
+    // The identifier of an application node, which picked up the index and processes it.
+    //
+    // Unset until a node picks the session.
+    NodeId picked_by = 2;
```
Comment (alexander-yevsyukov) [inline server/src/main/proto/spine/server/delivery/delivery.proto]: Please add the empty line.

File: server/src/main/proto/spine/server/delivery/delivery.proto [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+
+    // When the shard processed within the session was last picked by the node.
+    //
+    // This field is unset no nodes ever picked the session.
+    google.protobuf.Timestamp when_last_picked = 3;
```
Comment (alexander-yevsyukov) [inline server/src/main/proto/spine/server/delivery/delivery.proto]: Please add the empty line.

File: server/src/main/proto/spine/server/delivery/inbox.proto [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+import "google/protobuf/timestamp.proto";
+
+import "spine/client/entities.proto";
+import "spine/core/event.proto";
+import "spine/core/command.proto";
+import "spine/server/delivery/delivery.proto";
+
+
+// An identifier of an entity inbox.
+//
```
Comment (alexander-yevsyukov) [inline server/src/main/proto/spine/server/delivery/inbox.proto]: Please remove the empty line.

File: server/src/main/proto/spine/server/delivery/inbox.proto [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+
+    // The ID of the entity.
+    spine.client.EntityId entity_id = 1;
+
+    // The type URL of the entity state.
+    string type_url = 2;
+}
+
+// A label telling how to dispatch a message, residing in the `Inbox`.
+//
```
Comment (alexander-yevsyukov) [inline server/src/main/proto/spine/server/delivery/inbox.proto]: Please remove the empty line.

File: server/src/main/proto/spine/server/delivery/inbox.proto [hunk 30]
```diff
@@ -0,0 +1,136 @@
...
+
+    // Pass an event message to a `@React` or `@Command`-annotated event handler.
+    REACT_UPON_EVENT = 4;
+
+    // Pass an event message to a `@Apply(allowImport = "true")`-annotated event handler.
+    IMPORT_EVENT = 5;
+}
+
+// A status of the message in the `Inbox`.
+//
```
Comment (alexander-yevsyukov) [inline server/src/main/proto/spine/server/delivery/inbox.proto]: Please remove the empty line.

File: server/src/main/proto/spine/server/delivery/inbox.proto [hunk 31]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+// Identifies the original signal message that is packed inside the `InboxMessage`.
+message InboxSignalId {
+
+    // The value of the identifier of the source signal message, that has been put into the `Inbox`.
+    //
+    // E.g. if an instance of `Event` is put into the `Inbox`,
+    // this field has a value of `Event.id.value`.
+    string value = 1 [(required) = true];
```
Comment (alexander-yevsyukov) [inline server/src/main/proto/spine/server/delivery/inbox.proto]: Please add an empty line.

File: server/src/main/proto/spine/server/delivery/inbox.proto [hunk 32]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+
+// A message in the inbox.
+//
```
Comment (alexander-yevsyukov) [inline server/src/main/proto/spine/server/delivery/inbox.proto]: Please remove the empty line.

File: server/src/main/proto/spine/server/server_environment.proto [hunk 33]
```diff
@@ -25,3 +25,10 @@ enum DeploymentType {
     // Running on an App Engine cloud infrastructure.
     APPENGINE_CLOUD = 2;
 }
+
+// An identifier of a server node of the application.
+//
```
Comment (alexander-yevsyukov) [inline server/src/main/proto/spine/server/server_environment.proto]: Please remove the empty line.

**Pull request merged**